### PR TITLE
interfaces: relax rules for mount-control `what` for functionfs

### PR DESCRIPTION
**Pull request by mvo5**
The `functionfs` `what` rules are too strict. With functionfs pretty much anything can be in the `what` line so best to not bother, especially since functionfs requires a explicit mention in the plug declaration.

### CODE REVIEW

**Review by Meulengracht — APPROVED** [review]
Looks nice, +1! Would this need a security review? (Asking because curious)

**Review by MiguelPires — APPROVED** [review]
LGTM

**Comment by mvo5** [comment]
> Looks nice, +1! Would this need a security review? (Asking because curious)

I added Samuele and will let him decide if we need security or not.

File: interfaces/builtin/mount_control.go [hunk 1]
```diff
@@ -271,7 +271,14 @@ func enumerateMounts(plug interfaces.Attrer, fn func(mountInfo *MountInfo) error
 	return nil
 }
 
-func validateWhatAttr(what string) error {
+func validateWhatAttr(mountInfo *MountInfo) error {
+	// with "functionfs" the "what" can essentially be anything, see
```
Comment (pedronis) [inline interfaces/builtin/mount_control.go:275]: s/anything/any string/ ?

**Review by pedronis — COMMENTED** [review]
+1

**Review by pedronis — DISMISSED** [review]

File: interfaces/builtin/mount_control.go [hunk 2]
```diff
@@ -271,7 +271,14 @@ func enumerateMounts(plug interfaces.Attrer, fn func(mountInfo *MountInfo) error
...
+	// https://www.kernel.org/doc/html/latest/usb/functionfs.html
+	if len(mountInfo.types) == 1 && mountInfo.types[0] == "functionfs" {
+		return nil
```
Comment (pedronis) [inline interfaces/builtin/mount_control.go:278]: we should probably to be safe apply ValidateNoAppArmorRegexp on what I think

**Review by pedronis — COMMENTED** [review]
one comment

File: interfaces/builtin/mount_control.go [hunk 3]
```diff
@@ -271,7 +272,15 @@ func enumerateMounts(plug interfaces.Attrer, fn func(mountInfo *MountInfo) error
...
 }
 
-func validateWhatAttr(what string) error {
+func validateWhatAttr(mountInfo *MountInfo) error {
+	what := mountInfo.what
+
+	// with "functionfs" the "what" can essentially be anything, see
+	// https://www.kernel.org/doc/html/latest/usb/functionfs.html
+	if len(mountInfo.types) == 1 && mountInfo.types[0] == "functionfs" {
+		return apparmor_sandbox.ValidateNoAppArmorRegexp(what)
```
Comment (pedronis) [inline interfaces/builtin/mount_control.go:281]: this needs to be wrapped in some contextual info I think, see the test, the error looks different from the other

**Review by pedronis — APPROVED** [review]
thanks, remark about error

**Pull request merged**